File: examples/chapter7/figure_7-4_table_7-1_oliveoil.py
import warnings
warnings.filterwarnings("ignore")

import numpy as np
from sklearn.cross_decomposition import PLSRegression
from sklearn.metrics import confusion_matrix
from sklearn.linear_model import LinearRegression
from tabulate import tabulate
from sliced import SlicedInverseRegression
import matplotlib.pyplot as plt

from lib.lineal_with_R_more_classes import lineal_junto_small_R_more_classes

## TABLE 7.1
np.random.seed(180)

FITR_Spectra_olive_oils = np.genfromtxt("data/chapter7/FTIR_Spectra_olive_oils.csv", delimiter=",")

Y = (FITR_Spectra_olive_oils[1, 1:121].T - 1).reshape(-1, 1)
X = (FITR_Spectra_olive_oils[3:571, 1:121].T)

# Choose d with k=5 folds
Dmax = 33
K = 5

linear_error_pfc = [-1]*Dmax
linear_error_pls = [-1]*Dmax
linear_error_iso = [-1]*Dmax

for d in range(1, Dmax+1):

    aux = lineal_junto_small_R_more_classes(X, Y, K, d)

    li = aux['predict_linear_pls']
    li_pfc = aux['predict_linear_pfc']
    li_iso = aux['predict_linear_iso']

    linear_error_pls[d-1] = np.mean((li - Y)**2)
    linear_error_pfc[d-1] = np.mean((li_pfc - Y)**2)
    linear_error_iso[d-1] = np.mean((li_iso - Y)**2)

d_pls = np.argmin(linear_error_pls)+1
d_pfc = np.argmin(linear_error_pfc)+1
d_iso = np.argmin(linear_error_iso)+1

aux_pls = lineal_junto_small_R_more_classes(X, Y, X.shape[0], d_pls, loocv=True)['predict_linear_pls']
aux_pfc = lineal_junto_small_R_more_classes(X, Y, X.shape[0], d_pls, loocv=True)['predict_linear_pfc']
aux_iso = lineal_junto_small_R_more_classes(X, Y, X.shape[0], d_pls, loocv=True)['predict_linear_iso']

cm = confusion_matrix(y_true=Y, y_pred=aux_pls-1)
plserr = 100*(cm.diagonal().sum() / cm.sum())

cm = confusion_matrix(y_true=Y, y_pred=aux_iso-1)
isoerr = 100*(cm.diagonal().sum() / cm.sum())

cm = confusion_matrix(y_true=Y, y_pred=aux_pfc-1)
pls_pfcerr = 100*(cm.diagonal().sum() / cm.sum())

table = {
    "Dataset": ["Coffee"],
    "PLS": [f"{100-plserr:.2f}"],
    "ISO": [f"{100-isoerr:.2f}"],
    "PLS+PFC": [f"{100-pls_pfcerr:.2f}"]
}

print("Table 7.1")
print(tabulate(table, headers=table.keys()))
print()

## Figure 7.4 a
# PLS
d = d_pls
mod = PLSRegression(n_components=d, scale=False)
mod.fit(X, Y)
AUX = mod.x_loadings_

proj_data = X @ AUX

plt.title("Figure 7.4.a: PLS Projections")
plt.xlabel("First PLS Projection")
plt.ylabel("Second PLS Projection")
plt.scatter(x=proj_data[(Y==0).flatten()][:, 0], y=proj_data[(Y==0).flatten()][:, 1], label="Greece", marker="o")
plt.scatter(x=proj_data[(Y==1).flatten()][:, 0], y=proj_data[(Y==1).flatten()][:, 1], label="Italy", marker="^")
plt.scatter(x=proj_data[(Y==2).flatten()][:, 0], y=proj_data[(Y==2).flatten()][:, 1], label="Portugal", marker="s")
plt.scatter(x=proj_data[(Y==3).flatten()][:, 0], y=proj_data[(Y==3).flatten()][:, 1], label="Spain", marker="+")
plt.legend(loc='best')
plt.show()


## Figure 7.4 b
# PFC
d=d_pfc
mod = PLSRegression(n_components=d, scale=False)
mod.fit(X, Y)
AUX = mod.x_loadings_

proj_data = X @ AUX

s0 = SlicedInverseRegression()
s0.fit(X=proj_data, y=Y.flatten())
vecs = s0.directions_[np.argsort(s0.eigenvalues_)[::-1][:2]].T

proj_data_pfc = proj_data @ vecs

plt.title("Figure 7.4.b: PLS+PFC Projections")
plt.xlabel("First PFC Projection")
plt.ylabel("Second PFC Projection")
plt.scatter(x=proj_data_pfc[(Y==0).flatten()][:, 0], y=proj_data_pfc[(Y==0).flatten()][:, 1], label="Greece", marker="o")
plt.scatter(x=proj_data_pfc[(Y==1).flatten()][:, 0], y=proj_data_pfc[(Y==1).flatten()][:, 1], label="Italy", marker="^")
plt.scatter(x=proj_data_pfc[(Y==2).flatten()][:, 0], y=proj_data_pfc[(Y==2).flatten()][:, 1], label="Portugal", marker="s")
plt.scatter(x=proj_data_pfc[(Y==3).flatten()][:, 0], y=proj_data_pfc[(Y==3).flatten()][:, 1], label="Spain", marker="+")
plt.legend(loc='best')
plt.show()

# Figure 7.4.c
d=Dmax
# ISO
reg = LinearRegression()
# Inverse regression
reg.fit(Y, X)
residuals = X - reg.predict(Y) 

_, evecs = np.linalg.eigh(np.cov(residuals, rowvar=False))

evecs = evecs[:, ::-1]
evecs = evecs[:, :d]

proj_data_iso = X @ evecs

plt.title("Figure 7.4.c: Isotropic (ISO) Projections")
plt.xlabel("First ISO Projection")
plt.ylabel("Second ISO Projection")
plt.scatter(x=proj_data_iso[(Y==0).flatten()][:, 0], y=proj_data_iso[(Y==0).flatten()][:, 1], label="Greece", marker="o")
plt.scatter(x=proj_data_iso[(Y==1).flatten()][:, 0], y=proj_data_iso[(Y==1).flatten()][:, 1], label="Italy", marker="^")
plt.scatter(x=proj_data_iso[(Y==2).flatten()][:, 0], y=proj_data_iso[(Y==2).flatten()][:, 1], label="Portugal", marker="s")
plt.scatter(x=proj_data_iso[(Y==3).flatten()][:, 0], y=proj_data_iso[(Y==3).flatten()][:, 1], label="Spain", marker="+")
plt.legend(loc='best')
plt.show()

File: examples/chapter7/figure_7-4_table_7-1_oliveoil.R
rm(list=ls())
library(caret)
source("examples/chapter7/lib/lineal_with_R_more_classes.R")
source("examples/chapter7/lib/pls_matrices.R")
library(readr)
library(mltools)
library(data.table)


library(caret)

library(readr)
FTIR_Spectra_olive_oils <- read_csv("data/chapter7/FTIR_Spectra_olive_oils.csv")


Y=as.factor(t(FTIR_Spectra_olive_oils[2,2:121]))



Y_numeric=as.numeric(Y)
Y=matrix(Y_numeric,ncol=1)


xx=t(FTIR_Spectra_olive_oils[3:570,2:121])

X=matrix(0,ncol=ncol(xx),nrow=nrow(xx))
for (i in 1:ncol(xx)){
  for (j in 1:nrow(xx)){
    X[j,i]=as.numeric(xx[j,i])
  }
}


Dmax=33 

K=5


### Use this loop to choose d

error_lineal_PFC=NULL
error_lineal_pls=NULL
error_lineal_iso=NULL
set.seed(1)
for (d in 1:Dmax){

  aux=lineal.junto.small.R.more.classes(X,Y,K,d)
  
  li=as.numeric(aux$predict_lineal_pls)
  li_PFC=as.numeric(aux$predict_lineal_PFC)
  li_iso=as.numeric(aux$predict_lineal_iso)
  
  YY=as.numeric(Y)
  error_lineal_pls[d]=mean((li-YY)^2)
  error_lineal_PFC[d]=mean((li_PFC-YY)^2)
  error_lineal_iso[d]=mean((li_iso-YY)^2)
  
  
}






d_pls=which(error_lineal_pls==min(error_lineal_pls))
d_PFC=which(error_lineal_PFC==min(error_lineal_PFC))
d_iso=which(error_lineal_iso==min(error_lineal_iso))
 

aux_pls=lineal.junto.small.R.more.classes(X,Y,nrow(X),d_pls[1])$predict_lineal_pls
aux_PFC=lineal.junto.small.R.more.classes(X,Y,nrow(X),d_PFC[1])$predict_lineal_PFC
aux_iso=lineal.junto.small.R.more.classes(X,Y,nrow(X),d_iso[1])$predict_lineal_iso



# For table 7.1, the second row, first 3 columns

col1 = confusionMatrix(as.factor(aux_pls),as.factor(Y))$overall[1]
col2 = confusionMatrix(as.factor(aux_PFC),as.factor(Y ))$overall[1]
col3 = confusionMatrix(as.factor(aux_iso),as.factor(Y ))$overall[1]

cat("Table 7.1\nDataset   | PLS  | ISO | PFC\nOlive Oil |", round(col1,2), "|", round(col2,2), "|", round(col3,2), "\n")

### Figure 7.4

#######pls 
AUX=pls2_nipals(X ,Y ,d)$P[,1:d_pls]

datos_proy=X%*%AUX


grayPalette <- c("#555555", "#AAAAAA", "#717171", "#8D8D8D")


dfp_proj <- as.data.frame(datos_proy)
origins <- c("Greece", "Italy", "Portugal", "Spain")
dfp_proj$origins <- origins[Y] 

ggplot(dfp_proj, aes(x = V1, y = V2, shape = origins, color = origins)) + 
  geom_point(alpha = 0.7, size = 4) +
  theme_bw() +
  scale_color_manual(values=grayPalette) +
  labs(x = "First PLS projection",
       y = "Second PLS projection") +
  theme(panel.grid = element_blank(), 
        legend.position = c(0.9, 0.9), 
        legend.title = element_blank(),
        axis.title = element_text(size=16),
        legend.text = element_text(size = 13)
  ) 




dfp_proj <- as.data.frame(datos_proy)
origins <- c("Greece", "Italy", "Portugal", "Spain")
dfp_proj$origins <- origins[Y] 

ggplot(dfp_proj, aes(x = V1, y = V2, shape = origins, color = origins)) + 
  geom_point(alpha = 0.7, size = 4) +
  theme_bw() +
  labs(x = "First PLS projection",
       y = "Second PLS projection") +
  theme(panel.grid = element_blank(), 
        legend.position = c(0.9, 0.9), 
        legend.title = element_blank(),
        axis.title = element_text(size=16),
        legend.text = element_text(size = 13)
  ) 





###### iso

reg=lm(X ~Y )
vec=eigen(cov(reg$residuals))$vectors[,1:d_iso]
datos_proy_iso=X%*%vec


dfp_proj_iso <- as.data.frame(datos_proy_iso)
origins <- c("Greece", "Italy", "Portugal", "Spain")
dfp_proj_iso$origins <- origins[Y] 

ggplot(dfp_proj_iso, aes(x = V1, y = V2, shape = origins, color = origins)) + 
  geom_point(alpha = 0.7, size = 4) +
  theme_bw() +
  labs(x = "First ISO projection",
       y = "Second ISO projection") +
  scale_color_manual(values=grayPalette) +
  theme(panel.grid = element_blank(), 
        legend.position = c(0.9, 0.9), 
        legend.title = element_blank(),
        axis.title = element_text(size=16),
        legend.text = element_text(size = 13)
  ) 





dfp_proj_iso <- as.data.frame(datos_proy_iso)
origins <- c("Greece", "Italy", "Portugal", "Spain")
dfp_proj_iso$origins <- origins[Y] 

ggplot(dfp_proj_iso, aes(x = V1, y = V2, shape = origins, color = origins)) + 
  geom_point(alpha = 0.7, size = 4) +
  theme_bw() +
  labs(x = "First ISO projection",
       y = "Second ISO projection") +
  theme(panel.grid = element_blank(), 
        legend.position = c(0.9, 0.9), 
        legend.title = element_blank(),
        axis.title = element_text(size=16),
        legend.text = element_text(size = 13)
  ) 




#### pfc



AUX=pls2_nipals(X ,Y ,d)$P[,1:d_PFC]

datos_proy=X%*%AUX







s0 <- dr(Y~datos_proy)
datos_proy_pfc=datos_proy%*%s0$evectors[,1:2] 




dfp_proj_pfc <- as.data.frame(datos_proy_pfc)
origins <- c("Greece", "Italy", "Portugal", "Spain")
dfp_proj_pfc$origins <- origins[Y] 

ggplot(dfp_proj_pfc, aes(x = Dir1, y = Dir2, shape = origins, color = origins)) + 
  geom_point(alpha = 0.7, size = 4) +
  theme_bw() +
  labs(x = "First PFC projection",
       y = "Second PFC projection") +
  scale_color_manual(values=grayPalette) +
  theme(panel.grid = element_blank(), 
        legend.position = c(0.9, 0.9), 
        legend.title = element_blank(),
        axis.title = element_text(size=16),
        legend.text = element_text(size = 13)
  ) 



dfp_proj_pfc <- as.data.frame(datos_proy_pfc)
origins <- c("Greece", "Italy", "Portugal", "Spain")
dfp_proj_pfc$origins <- origins[Y] 

ggplot(dfp_proj_pfc, aes(x = Dir1, y = Dir2, shape = origins, color = origins)) + 
  geom_point(alpha = 0.7, size = 4) +
  theme_bw() +
  labs(x = "First PFC projection",
       y = "Second PFC projection") +
  theme(panel.grid = element_blank(), 
        legend.position = c(0.9, 0.9), 
        legend.title = element_blank(),
        axis.title = element_text(size=16),
        legend.text = element_text(size = 13)
  ) 






 